File: src/main/java/org/tinhol/mapper/api/Mapping.java
package org.tinhol.mapper.api;

import org.tinhol.mapper.Main;

import java.util.List;

public final class Mapping {

    private Mapping() {
    }

    public static Trivial trivial(String name) {
        return new Trivial(name);
    }

    public static Simple simple(String from, String to) {
        return new Simple(from, to);
    }

    public static Transform transform(String from, String to, Transformation transformation) {
        return new Transform(from, to, transformation);
    }

    public static Value value(Object value, String to) {
        return new Value(value, to);
    }

    public static <T extends ReadCommand, S> Main<T, S> mapping(Reader<T> reader, List<Op> mapping, TargetFactory<S> targetFactory) {
        return new Main<>(reader, mapping, targetFactory);
    }
}

File: src/main/kotlin/org/tinhol/mapper/api/Mapping.kt
package org.tinhol.mapper.api

sealed class Op
data class Trivial(val name: String) : Op()
data class Simple(val from: String, val to: String) : Op()
data class Transform(val from: String, val to: String, val transform : Transformation) : Op()
data class Value(val value: Any, val to: String) : Op()